Fixes #4530: Removes extra lines and spaces at the end of string in RTE

WickedBrat opened the pull request:
Fixes #4530 

I've fixed the issue with implemention using jQuery element.

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,22 @@ oppia.factory('RteHelperService', [
+        var textElt = elt[0].childNodes;
+        for (var i = textElt.length - 1; i > 0; i--) {
+          if (textElt[i].localName === 'p') {
+            test = textElt[ i].innerHTML;
```
This is still treating HTML content as a string. Instead the element should be analyzed with regards to child elements, and only "text nodes" in the DOM should be treated as strings.

E.g. you could look for all child elements; if all are br and all the text nodes are full of spaces only, then you can remove the paragraph DOM node.

seanlip replied:
Also please use more clear variable names; `test` isn't very helpful. In this context it should be something like `eltHtml` (but, as per the previous comment, I recommend a different approach instead).

WickedBrat replied:
@seanlip Done changes

seanlip replied:
Doesn't look done to me -- see https://github.com/oppia/oppia/pull/4799/files

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,19 @@ oppia.factory('RteHelperService', [
+        var textElt = elt[0].childNodes;
+        for (var i = textElt.length - 1; i > 0; i--) {
+          if (textElt[i].localName === 'p') {
+            if (textElt[i].childNodes[0].nodeValue === null ||
```
Unless I'm missing something, this feels a bit careless:

- What if there's 0 child nodes?
- What if there's more than one child node?
- What happened to handling `<br>` tags?

Make sure to consider all the possible cases exhaustively.

seanlip replied:
- For the first case, how many child nodes does `<p></p>` have?
- But you never check for "more than one tag", if I'm reading the code correctly.
- What if the paragraph contains a br tag, and then some other stuff?

WickedBrat replied:
- Here, the for loop is for number of tags in ```<p></p>```. It's the part which contains the text in each line.
- We just have to check for one tag, If there are more than one tags in the paragraph, it means that it doesn't need to check further.
- If there a ```<br>``` then it goes to next line and text is there, if there is some text and new line then it's new para.

seanlip replied:
- I don't think so. textElt[i] represents a "p" node and textElt[i].childNodes represents its children. children[0] takes the first child, but this assumes that there _is_ a first child. That assumption is what I'm asking about. Isn't it possible for the "p" node to contain no children?
- That is not correct either, I think. If the first node is null or just whitespace then your code would remove the paragraph even if the paragraph has later child nodes that aren't just whitespace. That's because the code is examining only the first child node.
- I don't think I understand your response, sorry. I'm talking about the possible case where a paragraph might contain br tags and some other things. This is a special case of point 2. (Also, there might be a case where br tags appear outside paragraphs; your code will not catch this because it is only looking at p tags. I wanted to check whether this is intentional.)

WickedBrat replied:
- Yes I get this, I've fixed it
- I'm sorry I couldn't generate this possibilty. Attaching pics for reference: 
![screenshot from 2018-03-19 00-12-17](https://user-images.githubusercontent.com/20617436/37571165-0c3b49ea-2b1f-11e8-867a-156f19958fe3.png)
![screenshot from 2018-03-19 02-40-53](https://user-images.githubusercontent.com/20617436/37571166-0c6cc448-2b1f-11e8-84fa-092070213a0b.png)
- Yes that's the technique I'm using for solving the issue. It's expected.

seanlip replied:
Hi @WickedBrat -- re your second/third point, the fix should work in the general case (so assume that any valid HTML string can be produced). This is because we are occasionally changing the framework for the RTE (e.g. there is a GSoC project to move it to CKEditor instead of textAngular) and so this code needs to be future-proof. (So try to analyze and solve the problem cleanly in the general case, even though you may not be able to repro it with the current text editor.)

Also, please make sure to add tests that exercise all the various corner cases!

WickedBrat replied:
Hi @seanlip I fixed the code as requested, but the frontend tests fail here: 
![screenshot from 2018-03-19 16-01-59](https://user-images.githubusercontent.com/20617436/37590914-c12b6bbc-2b8f-11e8-8647-322b8dbc7fdf.png)

I checked only the 'p' tags for the sane reason.

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,21 @@ oppia.factory('RteHelperService', [
+        var textElt = elt[0].childNodes;
+        for (var i = textElt.length - 1; i > 0; i--) {
+          if (textElt[i].localName === 'p') {
+            if (textElt[i].childNodes[0] === undefined) {
```
textElt[i].childNodes.length === 0

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,21 @@ oppia.factory('RteHelperService', [
-        text = elt.html();
-        text = $filter('removeExtraLines')(text);
-        return text;
+        var textElt = elt[0].childNodes;
```
This code should be wrapped in a method and tested with Karma tests covering all the special cases.

WickedBrat replied:
Done

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,21 @@ oppia.factory('RteHelperService', [
+            if (textElt[i].childNodes[0] === undefined) {
+              textElt[i].remove();
+            } else if (textElt[i].childNodes[0].nodeValue === null ||
+              $.trim(textElt[i].childNodes[0].nodeValue) === '') {
```
Why not just textElt[i].childNodes[0].nodeValue.trim()?

WickedBrat commented:
PTAL @seanlip

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -154,6 +154,7 @@ oppia.factory('RteHelperService', [
           return rte;
         }
 
+
```
Drop this newline. In general please check the "Files Changed" tab and remove spurious changes.

WickedBrat replied:
Sorry, I didn't notice this, fixed it.

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +182,29 @@ oppia.factory('RteHelperService', [
+          if (textElt[i].localName !== 'p') {
+            break;
+          }
+          if (textElt[i].childNodes.length === 0) {
```
I don't think you need lines 190-193, right?

WickedBrat replied:
Actually this is for the time when there is something like ```<p></p>```. So for this case I want to remove the tag right away and don't want to execute further. But if I remove these lines (190-193), the code will execute the for loop and give console error ```Cannot read property length of undefined```

seanlip replied:
I don't understand. Wouldn't j just initialize to 0 in that case?

WickedBrat replied:
Actually not, the value of j is ```textElt[i].childNodes.length```, for which it'll try to find childnodes of j, but will be undefined and accessing property of undefined will give a console error.

seanlip replied:
I still don't understand. This condition only triggers when textElt[i].childNodes.length = 0. So, if you say `j = textElt[i].childNodes.length` wouldn't j then be assigned the value 0?

seanlip commented on core/templates/dev/head/services/RteHelperServiceSpec.js:
```diff
@@ -0,0 +1,41 @@
+      .toEqual('<p>abc</p>');
+    expect(RteHelperService.convertRteToHtml('<p>abc</p><p>abc<br></p>'))
+      .toEqual('<p>abc</p><p>abc</p>');
+  });
```
Good tests. What about `<p>abc</p><br><br>`?

WickedBrat replied:
We actually don't need to worry about these cases. I looked into it and found that anything we type/insert into the text field gets wrapped around <p> tags, and that is ```rte``` that is passed into the function ```ConvertRteToHtml```. So if anything that is outside a <p> tag gets discarded automatically.

seanlip replied:
I thought I explained this in a previous review -- the code should work in the general case, since it will persist through multiple implementations of the RTE. So we do need to worry about that and similar cases.

WickedBrat replied:
In that case I'm updating the PR. Thanks!

WickedBrat commented:
@seanlip PTAL, I've updated the test cases and fixed the issues with code. Also I've updated  test case where ```<br>``` after div tag would cause an issue

seanlip commented on core/templates/dev/head/services/RteHelperService.js:
```diff
@@ -181,9 +181,26 @@ oppia.factory('RteHelperService', [
+            if (textElt[i].childNodes[j - 1].nodeValue === null ||
+              textElt[i].childNodes[j - 1].nodeValue.trim() === '') {
+              textElt[i].childNodes[j - 1].remove();
+            }
```
Suppose you have something along the lines of the following structure:

`<p> text <br> text <br> text <br> </p>`.

Is the intended result `<p> text <br> text <br> text </p>` or `<p> text text text </p>`? If the former then this code might not be doing what you want it to do.

Also, let's say the paragraph contains `<img>` tags, or other tags which have a null value. Do these get removed as well? I don't understand why we're checking for null nodes rather than e.g. just `<br>` and `<p>`.

Cases like these should be reflected in the tests.

WickedBrat replied:
> <p> text <br> text <br> text <br> </p>

Yes that's a good fix! I missed that.

>I don't understand why we're checking for null nodes rather than e.g. just <br> and <p>

Because of paragraph containing only spaces. I made a new fix, please check if it's fine

WickedBrat commented:
PTAL @seanlip
